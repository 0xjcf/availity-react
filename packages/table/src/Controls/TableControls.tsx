import React, { cloneElement, isValidElement, useEffect, useState } from 'react';
import { Form } from 'reactstrap';

type Props = {
  id?: string;
  disabled?: boolean;
  children?: React.ReactNode[] | React.ReactNode;
} & React.HTMLAttributes<HTMLElement>;

const TableControls = ({ id, disabled, children, ...rest }: Props): JSX.Element => {
  const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const childElements = Array.isArray(children) ? children : [children];
  return (
    <div id={id} className="table-controls" {...rest}>
      <Form inline>
        {childElements?.map((child: React.ReactNode, index: number) => {
          if (!isValidElement(child)) {
            return null;
          }

          return cloneElement(child, { disabled: isDisabled, key: `table_controls_${index}` });
        })}
      </Form>
    </div>
  );
};

export default TableControls;
