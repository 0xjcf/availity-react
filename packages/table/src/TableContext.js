import React, { useContext } from 'react';

export const TableContext = React.createContext();
export const useTableContext = () => useContext(TableContext);
