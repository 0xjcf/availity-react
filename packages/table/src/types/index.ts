export type { OnTableClickEvent } from './OnTableClickEvent';
export type { OnRowSelectedEvent } from './OnRowSelectedEvent';

export type { Row, Column, Cell, ExtendedTableHeader, CurrentTableState, TableInstance, RowProps } from './ReactTable';

export type { TableAction } from './TableAction';
export type { BulkRecordAction } from './BulkRecordAction';
export type { RecordAction } from './RecordAction';
export type { PrimaryRecordAction } from './PrimaryRecordAction';

export type { TableSort } from './TableSort';
export type { TableSortOption } from './TableSortOption';
