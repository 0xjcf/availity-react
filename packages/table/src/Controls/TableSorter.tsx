import React, { useEffect, useState } from 'react';
import {
  Button,
  ButtonDropdown,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  InputGroup,
  Label,
  UncontrolledTooltip,
} from 'reactstrap';
import Icon from '@availity/icon';
import find from 'lodash/find';
import { TableSort } from '../types/TableSort';
import { useTableContext } from '../TableContext';
import { CurrentTableState, IdType, TableInstance } from '../types/ReactTable';
import { TableSortOption } from '../types/TableSortOption';

type Props = {
  id?: string;
  disabled?: boolean;
  color?: string;
  onSort?: (sortBy: TableSort[]) => void;
  sortOptions?: TableSortOption[];
  autoGenerateSortOptions?: boolean;
} & React.HTMLAttributes<HTMLElement>;

const TableSorter = <T extends IdType>({ id, onSort, disabled, color, sortOptions, autoGenerateSortOptions, ...rest }: Props): JSX.Element | null => {
  const { sortableColumns, instance } = useTableContext();
  const { toggleSortBy, state } = instance as TableInstance<T>;

  const [isSortingDropdownOpen, setIsSortingDropdownOpen] = useState<boolean>(false);
  const [tableSort, setTableSort] = useState<TableSortOption | undefined>();
  const [tableSortDesc, setTableSortDesc] = useState<boolean>(true);
  const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);
  const [displayedSortOptions, setDisplayedSortOptions] = useState<TableSortOption[]>();

  useEffect(() => {
    if (sortableColumns) {
      let availableSortOptions: TableSortOption[] = [];
      if (autoGenerateSortOptions) {
        availableSortOptions = sortableColumns;
      }

      if (sortOptions) {
        availableSortOptions = [...availableSortOptions, ...sortOptions];
      }

      setDisplayedSortOptions(availableSortOptions);
    }
  }, [sortOptions, autoGenerateSortOptions, sortableColumns]);

  useEffect(() => {
    if (state && displayedSortOptions) {
      const { sortBy } = state as CurrentTableState;
      if (!sortBy) {
        return;
      }

      const currentSort = (sortBy as TableSort[])[0];
      if (!currentSort) {
        return;
      }

      const currentSortOption = find(displayedSortOptions, (option) => option.value === currentSort.id);
      if (currentSortOption) {
        setTableSort(currentSortOption);
        setTableSortDesc(currentSort.desc);
      }
    }
  }, [state, displayedSortOptions])

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const sort = async (sortBy: TableSortOption, isDesc: boolean) => {
    if (toggleSortBy) {
      toggleSortBy(sortBy.value, isDesc, false);
    }

    if (onSort) {
      onSort([{ id: sortBy.value, desc: isDesc }]);
    }
  };

  const toggleSortingDropdown = () => setIsSortingDropdownOpen(!isSortingDropdownOpen);

  const handleSort = async (sortOption: TableSortOption) => {
    await setTableSort(sortOption);

    let isDesc = tableSortDesc;
    if (sortOption.isDesc !== undefined) {
      isDesc = sortOption.isDesc;
      setTableSortDesc(isDesc);
    }

    sort(sortOption, isDesc);
  };

  const toggleSortDirection = async () => {
    const newSortDir = !tableSortDesc;
    await setTableSortDesc(newSortDir);

    if (!tableSort) {
      return;
    }

    sort(tableSort, newSortDir);
  };

  if (!displayedSortOptions) {
    return null;
  }

  return (
    <InputGroup
      data-testid="sorter_input_group"
      id={id}
      className="table-controls-sorting mb-2 me-sm-2 mb-sm-0"
      {...rest}
    >
      <Label className="me-sm-2 pr-1 pl-2">
        <strong>Sort By: </strong>
      </Label>
      <ButtonDropdown className="m1-1" isOpen={isSortingDropdownOpen} toggle={toggleSortingDropdown}>
        <DropdownToggle data-testid="sorter_toggle" disabled={isDisabled} color={color} caret>
          {tableSort && tableSort.label}
        </DropdownToggle>
        <DropdownMenu data-testid="sorter_menu" disabled={isDisabled} color={color}>
          {displayedSortOptions?.map((sortOption: TableSortOption) => (
            <DropdownItem
              data-testid={`sorter_menu_${sortOption.value}`}
              key={sortOption.value}
              onClick={() => handleSort(sortOption)}
            >
              {sortOption.label}
            </DropdownItem>
          ))}
        </DropdownMenu>
        <UncontrolledTooltip
          data-testid="tooltip_toggle_sort_dir"
          id="tooltip-toggle-sort-dir"
          placement="top"
          target="btn-toggle-sort-dir"
        >
          {tableSortDesc ? 'Descending' : 'Ascending'}
        </UncontrolledTooltip>
        <Button
          data-testid="btn_toggle_sort_dir"
          id="btn-toggle-sort-dir"
          disabled={isDisabled}
          color={color}
          onClick={async () => {
            await toggleSortDirection();
          }}
        >
          <Icon data-testid="btn_toggle_sort_dir_icon" name={tableSortDesc ? 'sort-alt-down' : 'sort-alt-up'} />
        </Button>
      </ButtonDropdown>
    </InputGroup>
  );
};

TableSorter.defaultProps = {
  autoGenerateSortOptions: true
}
export default TableSorter;
