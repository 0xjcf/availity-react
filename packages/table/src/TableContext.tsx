import React, { useContext } from 'react';
import { Column, TableSort, TableSortOption } from './types';
import { IdType } from './types/ReactTable';

export type AvTableContext = {
  scrollable?: boolean;
  setScrollable?: React.Dispatch<React.SetStateAction<boolean | undefined>>;
  sortable?: boolean;
  AdditionalContent?: React.ElementType;
  additionalContentProps?: Record<string, string | number | boolean | undefined | null>;
  toggleSelectAll?: () => void;
  sortBy?: TableSort;
  sortableColumns?: TableSortOption[];
  selectable?: boolean;
  hasCustomizableColumns: boolean;
  defaultColumns?: Column<IdType>[];
  onColumnsCustomized?: (hiddenColumnIds: string[], visibleColumnIds: []) => void;
  isCustomizingColumns?: boolean;
  setIsCustomizingColumns?: React.Dispatch<React.SetStateAction<boolean>>;
  minimumNumberOfColumns: number;
  onReset?: () => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  instance?: any;
};

export const TableContext = React.createContext<AvTableContext>({
  scrollable: false,
  sortable: false,
  hasCustomizableColumns: false,
  isCustomizingColumns: false,
  minimumNumberOfColumns: 3
});

export const useTableContext = (): AvTableContext => useContext(TableContext);
