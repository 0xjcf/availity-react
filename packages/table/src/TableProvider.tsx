import React, { useState } from 'react';
import { TableSortOption } from './types/TableSortOption';
import { TableContext } from './TableContext';
import { IdType, TableInstance } from './types/ReactTable';
import { CommonTableProps } from './Table';

export type TableProviderProps<T extends IdType> = {
  children: React.ReactChild | React.ReactChild[];
} & Partial<CommonTableProps<T>>;

const TableProvider = <T extends IdType>({ children }: TableProviderProps<T>): JSX.Element => {
  const [canScroll, setScrollable] = useState(false);
  const [canSelect, setSelectable] = useState(false);
  const [canSort, setSortable] = useState(false);

  const [tableInstance, setTableInstance] = useState<TableInstance<T>>();

  const [selectedRows, setSelectedRows] = useState<T[]>([]);
  const [sortableColumns, setSortableColumns] = useState<TableSortOption[]>([]);

  return (
    <TableContext.Provider
      value={{
        instance: tableInstance,
        setInstance: setTableInstance,

        scrollable: canScroll,
        setScrollable,

        sortable: canSort,
        setSortable,

        selectable: canSelect,
        setSelectable,

        selectedRows,
        setSelectedRows,

        sortableColumns,
        setSortableColumns,
      }}
    >
      {children}
    </TableContext.Provider>
  );
};

export default TableProvider;
