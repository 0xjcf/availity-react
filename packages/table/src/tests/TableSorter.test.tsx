import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import basicData from './data/basicData.json';
import TableControls from '../Controls/TableControls';
import TableProvider from '../TableProvider';
import { Column } from '../types/ReactTable';
import TableSorter from '../Controls/TableSorter';
import { TableContext } from '../TableContext';
import Table from '../Table';

const basicColumns = [
  {
    Header: 'First Name',
    accessor: 'first_name',
    defaultCanSort: true,
    disableSortBy: false,
  },
  {
    Header: 'Last Name',
    accessor: 'last_name',
    defaultCanSort: false,
    disableSortBy: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
  },
] as Column<Record<string, string | boolean | number>>[];

describe('TableSorter', () => {
  test('should render table Sorter', async () => {
    const { container, getByTestId } = render(
      <TableProvider>
        <TableControls>
          <TableSorter />
        </TableControls>
        <Table sortable data={basicData} columns={basicColumns} />
      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const sorterInputGrp = await waitFor(() => getByTestId('sorter_input_group'));
    expect(sorterInputGrp).not.toBeNull();
  });

  test('should render sort options in dropdown menu', async () => {
    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      state: {
        sortBy: [{ id: 'first_name', desc: true }],
      },
    };

    const { container, getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
          sortable: true,
          sortableColumns: [{ value: 'first_name', label: ' First name' }],
          selectedRows: [],
        }}
      >
        <TableControls>
          <TableSorter />
        </TableControls>
        <Table
          data={basicData}
          columns={basicColumns}
          initialState={{
            sortBy: [{ id: 'first_name', desc: true }],
          }}
        />
      </TableContext.Provider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const sorterMenu = await waitFor(() => getByTestId('sorter_menu'));
    expect(sorterMenu).not.toBeNull();

    const sorterMenuItem = await waitFor(() => getByTestId('sorter_menu_first_name'));
    expect(sorterMenuItem).not.toBeNull();
  });

  test('should show descending icon', async () => {
    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      state: {
        sortBy: [{ id: 'first_name', desc: true }],
      },
    };

    const { container, getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
          sortable: true,
          sortableColumns: [{ value: 'first_name', label: ' First name' }],
          selectedRows: [],
        }}
      >
        <TableControls>
          <TableSorter />
        </TableControls>
        <Table data={basicData} columns={basicColumns} />
      </TableContext.Provider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const sorterToggleButton = await waitFor(() => getByTestId('btn_toggle_sort_dir_icon'));
    expect(sorterToggleButton).not.toBeNull();
    expect(sorterToggleButton.className).toBe('icon icon-sort-alt-down');
  });

  test('should show ascending icon', async () => {
    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      state: {
        sortBy: [{ id: 'first_name', desc: false }],
      },
    };

    const { container, getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
          sortable: true,
          sortableColumns: [{ value: 'first_name', label: ' First name' }],
          selectedRows: [],
        }}
      >
        <TableControls>
          <TableSorter />
        </TableControls>
        <Table data={basicData} columns={basicColumns} />
      </TableContext.Provider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const sorterToggleButton = await waitFor(() => getByTestId('btn_toggle_sort_dir_icon'));
    expect(sorterToggleButton).not.toBeNull();
    expect(sorterToggleButton.className).toBe('icon icon-sort-alt-up');
  });

  test('should sort on dropdown select', async () => {
    const onSort = jest.fn();
    const toggleSortBy = jest.fn();

    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      state: {
        sortBy: [{ id: 'first_name', desc: false }],
      },
      toggleSortBy,
    };

    const { getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
          sortable: true,
          sortableColumns: [{ value: 'first_name', label: ' First name' }],
          selectedRows: [],
        }}
      >
        <TableControls>
          <TableSorter onSort={onSort} />
        </TableControls>
        <Table data={basicData} columns={basicColumns} />
      </TableContext.Provider>
    );

    const btnDropdownToggle = await waitFor(() => getByTestId('sorter_toggle'));
    expect(btnDropdownToggle).not.toBeNull();
    fireEvent.click(btnDropdownToggle);

    const btnAction = await waitFor(() => getByTestId('sorter_menu_first_name'));
    expect(btnAction).not.toBeNull();

    fireEvent.click(btnAction);

    await waitFor(() => {
      expect(onSort).toHaveBeenCalledTimes(1);
      expect(toggleSortBy).toHaveBeenCalledTimes(1);
    });
  });

  test('should populate sortOptions when provided', async () => {
    const onSort = jest.fn();
    const toggleSortBy = jest.fn();

    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      state: {
        sortBy: [{ id: 'first_name', desc: false }],
      },
      toggleSortBy,
    };

    const { getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
          sortable: true,
          sortableColumns: [{ value: 'first_name', label: ' First name' }],
          selectedRows: [],
        }}
      >
        <TableControls>
          <TableSorter onSort={onSort} sortOptions={[{ value: 'last_name', label: ' Last name' }]} />
        </TableControls>
        <Table data={basicData} columns={basicColumns} />
      </TableContext.Provider>
    );

    const btnDropdownToggle = await waitFor(() => getByTestId('sorter_toggle'));
    expect(btnDropdownToggle).not.toBeNull();
    fireEvent.click(btnDropdownToggle);

    const btnAction1 = await waitFor(() => getByTestId('sorter_menu_first_name'));
    expect(btnAction1).not.toBeNull();

    const btnAction2 = await waitFor(() => getByTestId('sorter_menu_last_name'));
    expect(btnAction2).not.toBeNull();
  });
});
