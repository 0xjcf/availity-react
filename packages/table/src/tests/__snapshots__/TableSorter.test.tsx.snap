// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSorter should render sort options in dropdown menu 1`] = `
<div>
  <div
    class="table-controls"
  >
    <form
      class="form-inline"
    >
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
             First name
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
               First name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
        >
          Tull
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
        >
          btull0@usatoday.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableSorter should render table Sorter 1`] = `
<div>
  <div
    class="table-controls"
  >
    <form
      class="form-inline"
    >
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              First Name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          style="cursor: pointer;"
          title="First Name"
        >
          First Name
          <i
            aria-hidden="true"
            class="icon icon-sort"
          />
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          style="cursor: pointer;"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableSorter should show ascending icon 1`] = `
<div>
  <div
    class="table-controls"
  >
    <form
      class="form-inline"
    >
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
             First name
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
               First name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-up"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableSorter should show descending icon 1`] = `
<div>
  <div
    class="table-controls"
  >
    <form
      class="form-inline"
    >
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
             First name
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
               First name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
